codegen-serde/src/test/kotlin/software/amazon/smithy/rust/codegen/serde/SerdeDecoratorTest.kt

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.serde
+
+import org.junit.jupiter.api.Test
+import software.amazon.smithy.rust.codegen.client.testutil.clientIntegrationTest
+import software.amazon.smithy.rust.codegen.core.rustlang.CargoDependency
+import software.amazon.smithy.rust.codegen.core.rustlang.CratesIo
+import software.amazon.smithy.rust.codegen.core.rustlang.RustType
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.testutil.IntegrationTestParams
+import software.amazon.smithy.rust.codegen.core.testutil.asSmithyModel
+import software.amazon.smithy.rust.codegen.core.testutil.integrationTest
+import software.amazon.smithy.rust.codegen.core.testutil.unitTest
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.server.smithy.testutil.serverIntegrationTest
+
+class SerdeDecoratorTest {
+    private val simpleModel =
+        """
+        namespace com.example
+        use smithy.rust#serde
+        use aws.protocols#awsJson1_0
+        use smithy.framework#ValidationException
+        @awsJson1_0
+        service HelloService {
+            operations: [SayHello, SayGoodBye],
+            version: "1"
+        }
+        @optionalAuth
+        operation SayHello {
+            input: TestInput
+            errors: [ValidationException]
+        }
+        @serde
+        structure TestInput {
```

Comment:
I think it would be worthwhile to have at least one recursive type in the test. I know it shows up in DDB's `AttributeValue` type so probably worth testing.

Reply:
done